<a href="https://colab.research.google.com/github/Maria-arroyo18/Aprendizaje-Supervisado-2025-/blob/main/Taller_Extra_Maria_Fernanda_Arroyo_Berrocal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#0. Cargamos las librerías y hacemos la conexión con Drive

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# 1. Cargamos los datos y verificamos la estructura del DataFrame

In [ ]:
path= '/content/drive/MyDrive/Aprendizaje/insurance.csv'
df= pd.read_csv(path)

In [ ]:
df.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

In [ ]:
df['region'].value_counts()

region
southeast    364
southwest    325
northwest    325
northeast    324
Name: count, dtype: int64

#2. Convertimos las variables categóricas a numéricas


In [ ]:
# Copiamos el DataFrame original
df_2 = df.copy()

# sex: male=1, female=0
df_2['sex'] = (df_2['sex'] == 'male').astype(int)

# smoker: yes=1, no=0
df_2['smoker'] = (df_2['smoker'] == 'yes').astype(int)

# Convierte la variable categórica 'region' en variables dummy (0/1) y elimina la original
df_2 = pd.get_dummies(df_2, columns=['region'], drop_first=False, dtype=int)

# Mostrar las primeras filas para verificar
df_2.head()

age  sex     bmi  children  smoker      charges  region_northeast  \
0   19    0  27.900         0       1  16884.92400                 0   
1   18    1  33.770         1       0   1725.55230                 0   
2   28    1  33.000         3       0   4449.46200                 0   
3   33    1  22.705         0       0  21984.47061                 0   
4   32    1  28.880         0       0   3866.85520                 0   

   region_northwest  region_southeast  region_southwest  
0                 0                 0                 1  
1                 0                 1                 0  
2                 0                 1                 0  
3                 1                 0                 0  
4                 1                 0                 0

In [ ]:
df_2.describe()

age          sex          bmi     children       smoker  \
count  1338.000000  1338.000000  1338.000000  1338.000000  1338.000000   
mean     39.207025     0.505232    30.663397     1.094918     0.204783   
std      14.049960     0.500160     6.098187     1.205493     0.403694   
min      18.000000     0.000000    15.960000     0.000000     0.000000   
25%      27.000000     0.000000    26.296250     0.000000     0.000000   
50%      39.000000     1.000000    30.400000     1.000000     0.000000   
75%      51.000000     1.000000    34.693750     2.000000     0.000000   
max      64.000000     1.000000    53.130000     5.000000     1.000000   

            charges  region_northeast  region_northwest  region_southeast  \
count   1338.000000       1338.000000       1338.000000       1338.000000   
mean   13270.422265          0.242152          0.242900          0.272048   
std    12110.011237          0.428546          0.428995          0.445181   
min     1121.873900          0.000000          0.000000          0.000000   
25%     4740.287150          0.000000          0.000000          0.000000   
50%     9382.033000          0.000000          0.000000          0.000000   
75%    16639.912515          0.000000          0.000000          1.000000   
max    63770.428010          1.000000          1.000000          1.000000   

       region_southwest  
count       1338.000000  
mean           0.242900  
std            0.428995  
min            0.000000  
25%            0.000000  
50%            0.000000  
75%            0.000000  
max            1.000000

#3. Entrenamos el modelo con los datos de entrenamiento

In [ ]:
train= df_2.sample(frac=0.7, random_state=42)
test= df_2.drop(train.index)

In [ ]:
train.head()

age  sex     bmi  children  smoker      charges  region_northeast  \
764    45    0  25.175         2       0   9095.06825                 1   
887    36    0  30.020         0       0   5272.17580                 0   
890    64    0  26.885         0       1  29330.98315                 0   
1293   46    1  25.745         3       0   9301.89355                 0   
259    19    1  31.920         0       1  33750.29180                 0   

      region_northwest  region_southeast  region_southwest  
764                  0                 0                 0  
887                  1                 0                 0  
890                  1                 0                 0  
1293                 1                 0                 0  
259                  1                 0                 0

In [ ]:
from sklearn.model_selection import train_test_split

X = df_2.drop('charges', axis=1)
y = df_2['charges']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [ ]:
from sklearn.linear_model import LinearRegression
import numpy as np

# Para usar las variables de la regresión, estas deben ser tipo array
X = np.array(X_train)
Y = np.array(y_train).reshape(-1, 1)

# Modelo de regresión
modelo = LinearRegression()
modelo.fit(X, Y)

# Ver coeficientes
print("Intercepto (a):", modelo.intercept_)
print("Pendiente (b):", modelo.coef_)

Intercepto (a): [-12970.02049227]
Pendiente (b): [[  261.29692414   104.81182299   348.90691516   424.11912829
  23628.36722236   596.05658924   109.12197876  -374.91224933
   -330.26631867]]


In [ ]:
df_train = pd.concat([y_train, X_train], axis=1)
df_train.head()

charges  age  sex     bmi  children  smoker  region_northeast  \
332  13429.03540   61    0  31.160         0       0                 0   
355  24603.04837   46    1  27.600         0       0                 0   
138  27322.73386   54    0  31.900         3       0                 0   
381  42303.69215   55    1  30.685         0       1                 1   
292  42112.23560   25    1  45.540         2       1                 0   

     region_northwest  region_southeast  region_southwest  
332                 1                 0                 0  
355                 0                 0                 1  
138                 0                 1                 0  
381                 0                 0                 0  
292                 0                 1                 0

In [ ]:
corr_matrix = df_train.corr()
corr_matrix

charges       age       sex       bmi  children    smoker  \
charges           1.000000  0.288804  0.040390  0.194264  0.079993  0.778989   
age               0.288804  1.000000 -0.013154  0.106366  0.061936 -0.046504   
sex               0.040390 -0.013154  1.000000  0.002457  0.018857  0.047368   
bmi               0.194264  0.106366  0.002457  1.000000 -0.004898 -0.003923   
children          0.079993  0.061936  0.018857 -0.004898  1.000000  0.023334   
smoker            0.778989 -0.046504  0.047368 -0.003923  0.023334  1.000000   
region_northeast  0.008044 -0.009283  0.016671 -0.156995  0.013583  0.010780   
region_northwest -0.038769  0.014390  0.042582 -0.125689  0.026750 -0.034933   
region_southeast  0.047119 -0.019003 -0.039575  0.250506 -0.021264  0.036953   
region_southwest -0.017719  0.014612 -0.019327  0.029865 -0.019050 -0.013916   

                  region_northeast  region_northwest  region_southeast  \
charges                   0.008044         -0.038769          0.047119   
age                      -0.009283          0.014390         -0.019003   
sex                       0.016671          0.042582         -0.039575   
bmi                      -0.156995         -0.125689          0.250506   
children                  0.013583          0.026750         -0.021264   
smoker                    0.010780         -0.034933          0.036953   
region_northeast          1.000000         -0.335138         -0.349670   
region_northwest         -0.335138          1.000000         -0.337948   
region_southeast         -0.349670         -0.337948          1.000000   
region_southwest         -0.328355         -0.317348         -0.331108   

                  region_southwest  
charges                  -0.017719  
age                       0.014612  
sex                      -0.019327  
bmi                       0.029865  
children                 -0.019050  
smoker                   -0.013916  
region_northeast         -0.328355  
region_northwest         -0.317348  
region_southeast         -0.331108  
region_southwest          1.000000

#4. Selección de características con Forward Selection y sklearn SequentialFeatureSelector

In [ ]:
import statsmodels.api as sm

def forward_selection(X_train, y_train):

    selected_features = []
    remaining_features = list(X_train.columns)
    best_aic = float('inf')

    while remaining_features:
        best_new_feature = None
        improved = False

        for feature in remaining_features:
            current_features = selected_features + [feature]
            X_temp = sm.add_constant(X_train[current_features])
            model = sm.OLS(y_train, X_temp).fit()
            current_aic = model.aic

            if current_aic < best_aic:
                best_aic = current_aic
                best_new_feature = feature
                improved = True

        if improved:
            selected_features.append(best_new_feature)
            remaining_features.remove(best_new_feature)
            print(f"Añadiendo {best_new_feature}. Current AIC: {best_aic:.2f}")
        else:
            break

    return selected_features

selected_features_forward = forward_selection(X_train, y_train)
print("\nVariables seleccionadas usando forward selection:", selected_features_forward)

Añadiendo smoker. Current AIC: 19385.47
Añadiendo age. Current AIC: 19093.83
Añadiendo bmi. Current AIC: 19004.22
Añadiendo children. Current AIC: 18999.47
Añadiendo region_northeast. Current AIC: 18998.62

Variables seleccionadas usando forward selection: ['smoker', 'age', 'bmi', 'children', 'region_northeast']


In [ ]:
from sklearn.linear_model import LinearRegression
from mlxtend.feature_selection import SequentialFeatureSelector as SFS
from sklearn.datasets import make_regression
import statsmodels.api as sm

# Modelo base
lr = LinearRegression()

# Forward selection
sfs = SFS(lr,
          k_features='best',   # Puede ser número fijo o 'best'
          forward=True,        # True = forward, False = backward
          floating=False,      # True para stepwise mixto
          scoring='r2',
          cv=5)

sfs = sfs.fit(X_train, y_train)

print("Variables seleccionadas usando SequentialFeatureSelector (Forward):", sfs.k_feature_names_)

# Crear y ajustar el modelo OLS con las variables seleccionadas
X_train_selected = sm.add_constant(X_train[list(sfs.k_feature_names_)])
model_sfs = sm.OLS(y_train, X_train_selected).fit()

# Resumen del modelo OLS
print("\nResumen del modelo OLS con variables seleccionadas por SequentialFeatureSelector:")
print(model_sfs.summary())

Variables seleccionadas usando SequentialFeatureSelector (Forward): ('age', 'bmi', 'children', 'smoker', 'region_northeast')

Resumen del modelo OLS con variables seleccionadas por SequentialFeatureSelector:
                            OLS Regression Results                            
Dep. Variable:                charges   R-squared:                       0.742
Model:                            OLS   Adj. R-squared:                  0.741
Method:                 Least Squares   F-statistic:                     535.3
Date:                Thu, 21 Aug 2025   Prob (F-statistic):          9.51e-271
Time:                        01:38:58   Log-Likelihood:                -9493.3
No. Observations:                 936   AIC:                         1.900e+04
Df Residuals:                     930   BIC:                         1.903e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
  

In [ ]:
from sklearn.metrics import mean_squared_error, mean_absolute_error
import numpy as np

# Predicciones con el set de prueba usando el modelo de entrenamiento con SFS features
X_test_selected = sm.add_constant(X_test[list(sfs.k_feature_names_)])
y_pred = model_sfs.predict(X_test_selected)

# RMSE
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print(f"RMSE: {rmse:.2f}")

# MAE
mae = mean_absolute_error(y_test, y_pred)
print(f"MAE: {mae:.2f}")

# MAPE
mape = np.mean(np.abs((y_test - y_pred) / y_test)) * 100
print(f"MAPE: {mape:.2f}%")

RMSE: 5819.36
MAE: 4151.31
MAPE: 43.72%


#5. Evaluación del rendimiento del modelo

**Análisis de métricas de desempeño**

**RMSE – Root Mean Squared Error**

En este modelo el RMSE es de 5819.36, lo que significa que, en promedio, las predicciones del modelo se desvían 5819.36 de los valores reales. Esto muestra que los errores de predicción son bastante altos, lo que puede indicar que el modelo aún no captura bien todas las relaciones de los datos.

**MAE – Mean Absolute Error**

En este caso, el MAE es de 4151.31, lo que indica que, en promedio, el modelo se equivoca en  4151.31 por predicción. Aunque es más bajo que el RMSE, sigue siendo un error considerable, lo que refuerza la idea de que el modelo no predice con alta precisión.

MAPE – Mean Absolute Percentage Error

En este modelo el MAPE es de 43.72%, que significa que, en promedio, el modelo se equivoca 43.7% con respecto al valor real de los cargos. Este valor indica un desempeño bajo, ya que un MAPE cercano al 10% se suele considerar aceptable en muchos contextos. En este dataset los valores de charges no son tan bajos, por lo que el valor refleja de forma confiable el desempeño.

**Conclusión**

El modelo presenta errores relativamente altos en todas las métricas (RMSE, MAE y MAPE). Esto significa que sus predicciones no son muy precisas para el conjunto de prueba.

In [ ]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import numpy as np
import statsmodels.api as sm

# Predicciones en el set de prueba con el modelo entrenado con SFS features
X_test_selected = sm.add_constant(X_test[list(sfs.k_feature_names_)])
y_pred = model_sfs.predict(X_test_selected)

# RMSE
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print(f"RMSE: {rmse:.2f}")

# MAE
mae = mean_absolute_error(y_test, y_pred)
print(f"MAE: {mae:.2f}")

# Calculate MAPE
# epsilon para evitar la division por cero
epsilon = 1e-10
mape = np.mean(np.abs((y_test - y_pred) / (y_test + epsilon))) * 100
print(f"MAPE: {mape:.2f}%")

# R2
r2 = r2_score(y_test, y_pred)
print(f"R2: {r2:.2f}")

# R2 ajustado
n = len(y_test) # numero de observaciones
p = X_test_selected.shape[1] - 1 # numero de predictores (excluyendo la constante)
adj_r2 = 1 - (1 - r2) * (n - 1) / (n - p - 1)
print(f"R2 ajustado: {adj_r2:.2f}")

RMSE: 5819.36
MAE: 4151.31
MAPE: 43.72%
R2: 0.77
R2 ajustado: 0.77


**R²**

En este caso, R² = 0.77, lo que significa que el modelo logra explicar aproximadamente el 77% de la variabilidad de los valores de charges. Este es un buen nivel de explicación, aunque el error absoluto todavía es alto.

**R² ajustado**

Aquí, el R² ajustado es de 0.77, casi igual al R² normal. Esto indica que las variables seleccionadas realmente aportan información y que el modelo no está sobreajustado.

**Conclusión general**

El modelo explica bien la variabilidad de los datos (R² = 0.77), pero sus errores de predicción (RMSE y MAE altos, MAPE cercano al 44%) muestran que todavía tiene dificultades para dar estimaciones precisas en valores individuales.
Esto puede ser por lo siguiente:

- La naturaleza de los datos (alta dispersión en charges).
- No linealidad en la relación entre predictores y la variable objetivo.